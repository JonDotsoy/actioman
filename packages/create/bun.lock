{
  "lockfileVersion": 1,
  "workspaces": {
    "": {
      "name": "create-actioman",
      "dependencies": {
        "@jondotsoy/flags": "^2.3.0",
      },
      "devDependencies": {
        "@types/bun": "latest",
      },
      "peerDependencies": {
        "typescript": "^5.0.0",
      },
    },
  },
  "packages": {
    "@jondotsoy/console-draw": ["@jondotsoy/console-draw@0.1.2", "", { "dependencies": { "@jondotsoy/style-text": "^0.9.1" }, "peerDependencies": { "prettier": "^3.3.3", "typescript": "^5.0.0" } }, "sha512-BBZGLd6RzjKxKcgD2OwD7v6GeKlf53Ysk59H+7+HvyGLcJZ7AJk8rOCq/yv813KbWV2teTCEBQXnnP5hZe/2CA=="],

    "@jondotsoy/flags": ["@jondotsoy/flags@2.3.0", "", { "dependencies": { "@jondotsoy/console-draw": "^0.1.2" } }, "sha512-vUMe4NdXKv8RxMXrKCJLsHBt3XQ8/YV4Fyr1I43vSLe5v0M8sdY3wJl1qdTgwRBZES2zMLoenxOUSNGE9r/5NA=="],

    "@jondotsoy/style-text": ["@jondotsoy/style-text@0.9.2", "", { "peerDependencies": { "typescript": "^5.0.0" } }, "sha512-6fonOzrANwrw+UIM4fivtDZLvuCN00yH0lCEYi14ln4j8fSHZokb6TLrqOu38GCqqQnZ2LuzJTQBWgv7jN6iVA=="],

    "@types/bun": ["@types/bun@1.2.2", "", { "dependencies": { "bun-types": "1.2.2" } }, "sha512-tr74gdku+AEDN5ergNiBnplr7hpDp3V1h7fqI2GcR/rsUaM39jpSeKH0TFibRvU0KwniRx5POgaYnaXbk0hU+w=="],

    "@types/node": ["@types/node@22.13.1", "", { "dependencies": { "undici-types": "~6.20.0" } }, "sha512-jK8uzQlrvXqEU91UxiK5J7pKHyzgnI1Qnl0QDHIgVGuolJhRb9EEl28Cj9b3rGR8B2lhFCtvIm5os8lFnO/1Ew=="],

    "@types/ws": ["@types/ws@8.5.14", "", { "dependencies": { "@types/node": "*" } }, "sha512-bd/YFLW+URhBzMXurx7lWByOu+xzU9+kb3RboOteXYDfW+tr+JZa99OyNmPINEGB/ahzKrEuc8rcv4gnpJmxTw=="],

    "bun-types": ["bun-types@1.2.2", "", { "dependencies": { "@types/node": "*", "@types/ws": "~8.5.10" } }, "sha512-RCbMH5elr9gjgDGDhkTTugA21XtJAy/9jkKe/G3WR2q17VPGhcquf9Sir6uay9iW+7P/BV0CAHA1XlHXMAVKHg=="],

    "prettier": ["prettier@3.4.2", "", { "bin": { "prettier": "bin/prettier.cjs" } }, "sha512-e9MewbtFo+Fevyuxn/4rrcDAaq0IYxPGLvObpQjiZBMAzB9IGmzlnG9RZy3FFas+eBMu2vA0CszMeduow5dIuQ=="],

    "typescript": ["typescript@5.7.3", "", { "bin": { "tsc": "bin/tsc", "tsserver": "bin/tsserver" } }, "sha512-84MVSjMEHP+FQRPy3pX9sTVV/INIex71s9TL2Gm5FG/WG1SqXeKyZ0k7/blY/4FdOzI12CBy1vGc4og/eus0fw=="],

    "undici-types": ["undici-types@6.20.0", "", {}, "sha512-Ny6QZ2Nju20vw1SRHe3d9jVu6gJ+4e3+MMpqu7pqE5HT6WsTSlce++GQmK5UXS8mzV8DSYHrQH+Xrf2jVcuKNg=="],
  }
}
