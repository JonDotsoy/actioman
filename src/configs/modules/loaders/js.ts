export default async (sourceLocation: string) => {
  const { default: def } = await import(sourceLocation);
  return def;
};
