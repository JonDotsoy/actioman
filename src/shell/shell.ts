import { spawn, spawnSync } from "child_process";

type ShellConstructorOutput = "codeStatus" | "text" | "pipe";

type ShellConstructorOptions = {
  cmd: string;
  cwd?: string;
  signal?: AbortSignal;
  env?: Record<string, string>;
  output: ShellConstructorOutput;
  silent?: boolean;
};

export class ShellConstructor<T> {
  private cmd: string;
  private cwdOptions?: string;
  private abortController: AbortController;
  private envOptions: Record<string, string> | undefined;
  private appendEnvsOptions: Record<string, string> = {};
  private shouldThrow = false;
  private status: "pending" | "running" | "fulfilled" | "rejected" = "pending";
  private _promise = Promise.withResolvers<any>();
  private output: ShellConstructorOutput;
  private silent: boolean | undefined;
  private stdoutReadableController?: ReadableStreamDefaultController<Uint8Array>;
  private stdoutReadable = new ReadableStream<Uint8Array>({
    start: (controller) => {
      this.stdoutReadableController = controller;
    },
  });
  private stdoutWritable = new WritableStream<Uint8Array>({
    write: (chunk) => {
      if (!this.silent) process.stdout.write(chunk);
      if (this.output === "text" || this.output === "pipe")
        this.stdoutReadableController?.enqueue(chunk);
    },
    close: () => {
      this.stdoutReadableController?.close();
    },
  });
  private stderrWritable = new WritableStream<Uint8Array>({
    write: (chunk) => {
      if (!this.silent) process.stderr.write(chunk);
    },
  });
  runned = Promise.withResolvers<{ pid: number | undefined }>();

  constructor(options: ShellConstructorOptions) {
    this.abortController = new AbortController();
    this.cmd = options.cmd;
    this.cwdOptions = options.cwd;
    this.envOptions = options.env;
    this.output = options.output;
    this.silent = options.silent;
    options.signal?.addEventListener("abort", (event) => {
      this.abortController.abort();
    });
  }

  cwd(newCwd: string) {
    this.cwdOptions = newCwd;
    return this;
  }

  appendEnvs(env: Record<string, string>) {
    this.appendEnvsOptions = {
      ...this.appendEnvsOptions,
      ...env,
    };
    return this;
  }

  env(newEnv: Record<string, string>) {
    this.envOptions = newEnv;
    return this;
  }

  throws(shouldThrow: boolean = true) {
    this.shouldThrow = shouldThrow;
    return this;
  }

  quiet(silent: boolean = true) {
    this.silent = silent;
    return this;
  }

  text(): ShellConstructor<string> {
    this.output = "text";
    return this.quiet() as any;
  }

  background() {
    this.output = "pipe";
    const pid = this.runned.promise.then((r) => r.pid);

    let closed = false;
    return {
      pid,
      exited: this.then(),
      stdout: this.stdoutReadable,
      stderr: this.stderrWritable,
      close: async () => {
        if (closed) return;
        this.abortController.abort();
        const p = await pid;
        try {
          spawnSync(`kill ${p}`);
        } catch {}
        closed = true;
      },
    };
  }

  async run() {
    const cmd = this.cmd;
    const cwd = this.cwdOptions;
    const env = {
      ...(this.envOptions ?? process.env),
      ...this.appendEnvsOptions,
    };
    const abortController = this.abortController;
    const throws = this.shouldThrow;
    const output = this.output;
    const stdoutWritableWriter = this.stdoutWritable.getWriter();
    const stderrWritableWriter = this.stderrWritable.getWriter();

    const subprocess = spawn("sh", ["-c", cmd], {
      signal: abortController.signal,
      stdio: ["pipe", "pipe", "pipe"],
      env: env,
      cwd: cwd,
    });

    subprocess.stdout.addListener("data", (chunk: Uint8Array) => {
      stdoutWritableWriter.write(chunk);
    });
    subprocess.stdout.addListener("close", () => {
      stdoutWritableWriter.close();
    });

    subprocess.stderr.addListener("data", (chunk: Uint8Array) => {
      stderrWritableWriter.write(chunk);
    });
    subprocess.stderr.addListener("close", () => {
      stderrWritableWriter.close();
    });

    const formatOutput = async (exitCode: number) => {
      if (output === "pipe") return exitCode;
      if (output === "codeStatus") return exitCode;
      if (output === "text") {
        const text: number[] = [];
        const reader = await this.stdoutReadable.getReader();
        let readableStreamDefaultReadResult: Bun.ReadableStreamDefaultReadResult<Uint8Array>;
        while ((readableStreamDefaultReadResult = await reader.read())) {
          if (readableStreamDefaultReadResult.done) break;
          text.push(...readableStreamDefaultReadResult.value);
        }
        return new TextDecoder().decode(new Uint8Array(text));
      }
    };

    subprocess.addListener("error", (err) => {
      this._promise.reject(err);
    });

    subprocess.addListener("close", (exitCode) => {
      if (typeof exitCode === "number") {
        if (exitCode !== 0 && !throws)
          return this._promise.reject(
            new Error(`exitCode: ${exitCode} cmd: ${cmd}`),
          );
        return this._promise.resolve(formatOutput(exitCode));
      }

      return this._promise.reject(
        new Error(`exitCode is not a number: ${exitCode}`),
      );
    });
  }

  then<TResult1 = T, TResult2 = never>(
    onfulfilled?:
      | ((value: T) => TResult1 | PromiseLike<TResult1>)
      | undefined
      | null,
    onrejected?:
      | ((reason: any) => TResult2 | PromiseLike<TResult2>)
      | undefined
      | null,
  ): Promise<TResult1 | TResult2> {
    if (this.status === "pending") this.run();
    return this._promise.promise.then(onfulfilled, onrejected);
  }
}

export const $ = (
  arr: TemplateStringsArray,
  ...args: string[]
): ShellConstructor<number> => {
  return new ShellConstructor({
    cmd: String.raw(arr, ...args),
    output: "codeStatus",
  });
};

$.cwd =
  (cwd: string) =>
  (arr: TemplateStringsArray, ...args: string[]) =>
    $(arr, ...args).cwd(cwd);
$.env =
  (env: Record<string, string>) =>
  (arr: TemplateStringsArray, ...args: string[]) =>
    $(arr, ...args).env(env);
