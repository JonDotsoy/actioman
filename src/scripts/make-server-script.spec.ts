import { describe, expect, it } from "bun:test";
import { $ } from "../shell/shell";
import { makeServerScript } from "./make-server-script.js";
import * as fs from "fs/promises";

describe("makeServerScript", () => {
  it("generates bootstrap script", async () => {
    const WORKSPACE_PATH = "./__tests__/__workspaces__/make-server-script/";

    await $`
      rm -rf $WORKSPACE_PATH
      mkdir -p $WORKSPACE_PATH
      echo "*" > $WORKSPACE_PATH/.gitignore
    `
      .cwd(new URL("./", import.meta.url).pathname)
      .appendEnvs({ WORKSPACE_PATH });

    await $`
      mkdir -p node_modules/actioman/lib/esm/http-router/
      touch node_modules/actioman/lib/esm/http-router/http-listener.js
      touch actions.js
    `.cwd(new URL(WORKSPACE_PATH, import.meta.url).pathname);

    const actionsPathRaw = await $`realpath actions.js`
      .cwd(new URL(WORKSPACE_PATH, import.meta.url).pathname)
      .text();
    const actionsPath = actionsPathRaw.trim();

    const res = await makeServerScript(
      new URL(WORKSPACE_PATH, import.meta.url).pathname,
      actionsPath,
    );

    expect(res.bootstrapLocation).toBeInstanceOf(URL);
    expect(
      await fs.readFile(new URL(res.bootstrapLocation), "utf-8"),
    ).toMatchSnapshot();
  });
});
