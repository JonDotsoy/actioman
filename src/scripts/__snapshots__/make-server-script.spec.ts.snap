// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`makeServerScript generates bootstrap script 1`] = `
"// @ts-nocheck
import { HTTPLister } from './../../../../../../../http-router/http-listener.js';
import { factory } from './../../../../../../../configs/modules/factory.js';
import { Configs } from './../../../../../../../configs/configs.js';
const PORT = process.env.PORT;
const HOST = process.env.HOST;
const bootstrap = async () => {
    const configs = Configs.fromModule(await factory.findOn(new URL('./../../../', import.meta.url).toString(), 'configs'));
    const httpLister = await HTTPLister.fromModule(await import('./../../../actions.js'), configs);
    const url = await httpLister.listen(PORT, HOST);
    console.log(\`Server running at \${ url.toString() }\`);
};
bootstrap().catch(console.error);"
`;
